Implemented HMM Archetype

gaffney2010 opened the pull request:
Moved HMMParams to its own file in archetypes folder; expanded to implement PSO algorithm; and added some tests.

In testing HMMParams, I changed:

1. Moved emission_probabilities randomize function to random_params function.
2. Fixed some bugs in parse_repr.

I added --algorithm to hmm_evolve, and ran algorithm based on this.

gaffney2010 committed:
Fixed Swap Nodes in mutate_rows per issue #41

Swapped based on state, row[0], then sorted.  As per our conversation in issue #41 thread.

gaffney2010 committed:
Updated for fixed swap in mutation

gaffney2010 committed:
Moved HMMParams class and implemented tests.

gaffney2010 committed:
Implemented PSO for HMM.

gaffney2010 committed:
Implemented PSO for HMM.

gaffney2010 committed:
Merge branch 'master' of https://github.com/gaffney2010/axelrod-dojo

gaffney2010 committed:
Deleted CSVs from bin.

Nikoleta-v3 commented:
Hello @gaffney2010, I am one of the contributors! 

You are getting an error in line `26` because you are trying to access `self.mutation_rate`. You have renamed this to be `self.mutation_probability`.

This is looking good. Would you mind adding some more tests? For example look at the integration tests here: https://github.com/Axelrod-Python/axelrod-dojo/tree/master/tests/integration

They explain how the prototypes can be used with the algorithms. So a file `test_hmm.py` could be added.

gaffney2010 commented:
HI @Nikoleta-v3.  Thanks for the input.

I renamed `mutation_rate` to `mutation_probability` because the genetic_algorithm looked for `mutation_probability` explicitly in the HMMParam keywords.  Maybe there's another way to work around this, but at any rate I thought this variable name should match other archetypes.  I missed the checks in my test file, but I've fixed that now so that it's `mutation_probability` everywhere.

I will work on the integration tests.

gaffney2010 committed:
Changed mutation_rate to mutation_probability in test.

Nikoleta-v3 commented:
Yes I agree with you on the name,  it should be `mutation_probability`.  Thank you!

marcharper commented on bin/hmm_evolve.py:
```diff
@@ -22,190 +22,15 @@
     --noise NOISE               Match noise [default: 0.00]
     --nmoran NMORAN             Moran Population Size, if Moran objective [default: 4]
     --states NUM_STATES         Number of FSM states [default: 5]
+    --algorithm ALGORITHM       Which algorithm to use (GA for genetic algorithm or PSO for
```
How about something like: EA (evolutionary algorithm) and PS (Particle Swarm)

drvinceknight reviewed (changes requested):
Could you add a sentence or two (feel free to add more) to the docs too please. Just something on this page loosely describing HMMs would do :+1:  http://axelrod-dojo.readthedocs.io/en/latest/background/genetic_algorithm.html

gaffney2010 commented:
I added some stuff to the documentation page.

I added some integration tests and fixed some errors as they came up.  In my last integration test, the unittest kept switching out while the temporary file was still open, causing errors.  I seem to have fixed this by adding the lines.
with open(self.temporary_file.name, "w") as f:
            pass
Maybe this is okay, because the other methods read from this file, which may have the same effect, but I'm honest not sure what's going on or why this fixes it.

gaffney2010 committed:
Added tests/docstrings to HMM, and fixed bugs in HMM.

drvinceknight commented on docs/background/genetic_algorithm.rst:
```diff
@@ -51,9 +51,44 @@ The crossover and mutation are implemented in the following way:
+- Crossover: this is done by taking a randomly selected number of rows from
+  one cooperation transition matrix and the rest from the other to form a target
+  cooperation transition matrix; then a different number of randomly selected
+  rows from one defection transition matrix and the restf from the other; and
```
`rest` (typo)

drvinceknight reviewed (changes requested):
Found a typo. :+1: Thanks for the docs :)

gaffney2010 committed:
Fixed typo

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,255 @@
+                   initial_state=initial_state,
+                   initial_action=initial_action)
+
+    def receive_vector(self, vector):
```
Can you combine the serialization / deserialization methods? The functionality of the `vector_to_instance` and `receive_vector` methods are covered by the repr class methods on the parameter class. I think the class methods contain all the logic needed to unpack the serialized parameters (so I don't think the quadratic equation is needed).

The classmethods could be improved of course so feel free to do so.

gaffney2010 replied:
Possible I'm missing something, but the parse_repr method parses a string and relies on the special character delimiters.  Here I need to parse a list of floats (what the PSO uses), and without the special characters, I don't see a natural way to delimit this...

marcharper replied:
Ok, I see how the vector functions are being used now with the PSO algorithm. Rather than have `receive_vector` and `vector_to_instance` as instance methods can you make these methods staticmethods / classmethods that return new objects, and name them something like `to_vector` and `from_vector`? Please also explain in the docstring why the additional representation exists (for the PSO algorithm).

Also, I think that `create_vector_bounds` isn't naturally a method of the parameter class, it need not care about an implementation detail of the PSO algorithm. Rather `create_vector_bounds` could be a standalone function (or a staticmethod) that just relies on `N` as a parameter or the size of the input vector to the PSO algorithm.

Finally, consider if it makes sense to supply `N` as a parameter to `vector_to_instance` rather than computing it from the quadratic equation. Are you sure that the chosen root is always the correct one?

gaffney2010 replied:
I will look into all of this.  I'm sure the chosen root is the correct one; the other must be negative.

gaffney2010 replied:
Alright.  I thought about this more, and I agree that what I had before was a sort-of hollow pass-through that didn't really need to live in the class.  On revisiting it and looking at what Gambler is trying to do, I think it makes sense to adjust `receive_vector` to not just store the passed vector in the class as a vector, but to actually deserialize there and store into the appropriate class variables.

When you do this, you can then use the the HMMParams instance like a normally-loaded instance of HMMParams.  You can see this where I changed my test file, instead of running the `read_vector` function to deserialize a `xopt` and put those parameters into an HMMParams class, I can read `xopt` (the vector) right into HMMParams using the `receive_vector` method.

Also if you do this, then the `vector_to_instance` method is already implemented as the `player` method.  [So I changed the name in PSO and Gambler implementations to `player` to be consistent.]  Finally, recognizing `receive_vector` and `create_vector_bounds` _as_ instance methods, you can just use `self.num_states` in these functions instead of using a quadratic.  [This is analogous to using `self.pattern` in the Gambler `create_vector_bounds.]

Then I made all these same changes to FSM.  [My implementation in HMM was modeled after FSM.  So I think if we fix in HMM, then we should fix in FSM.]

marcharper commented on tests/integration/test_hmm.py:
```diff
@@ -0,0 +1,419 @@
+        noise = 0
+        repetitions = 5
+        num_states = 2
+        opponents = [axl.ThueMorse(),
```
+1 for more integration tests but it's not necessary to single out the tricky to pickle strategies. It would be better to choose fast to run strategies such as TitForTat and WinStayLoseShift.

gaffney2010 replied:
I'll change this.

marcharper reviewed (changes requested):
Looking good overall, please see the comments on serialization.

gaffney2010 committed:
Changed test strategies from hard-to-pickle strategies to basic strategies.

gaffney2010 committed:
Merge https://github.com/gaffney2010/axelrod-dojo

gaffney2010 committed:
Made read_vector, create_vector_bounds, and vector_to_instance more instance-specific

gaffney2010 committed:
Made vector changes to FSM too.

gaffney2010 committed:
Updated test_hmm for vector_to_instance change.

gaffney2010 committed:
Updated test_gambler for vector_to_instance change.

gaffney2010 committed:
Fixed test_vector_to_instance in test_fsm.

marcharper commented on src/axelrod_dojo/archetypes/fsm.py:
```diff
@@ -140,34 +140,30 @@ def parse_repr(cls, s):
 
-        Finally, a probability to determine the player's first move."""
-
-        num_states = int((len(self.vector) - 1) / 4)
```
Let's use `//` instead of `int` (since we only support Python 3+)

gaffney2010 replied:
This code's been removed I think.

marcharper commented on src/axelrod_dojo/archetypes/fsm.py:
```diff
@@ -140,34 +140,30 @@ def parse_repr(cls, s):
-
-        num_states = int((len(self.vector) - 1) / 4)
-        state_scale = self.vector[:num_states * 2]
-        next_states = [int(s * (num_states - 1)) for s in state_scale]
```
I think the `int` is unnecessary here as well (all value should be integers)

marcharper commented on src/axelrod_dojo/archetypes/fsm.py:
```diff
@@ -140,34 +140,30 @@ def parse_repr(cls, s):
+        Finally, a probability to determine the player's first move.
+        """
+        state_scale = vector[:self.num_states * 2]
+        next_states = [int(s * (self.num_states - 1)) for s in state_scale]
```
same comment on `int`

gaffney2010 replied:
s should be a decimal in fact.

The PSO doesn't totally make for FSM...  This line converts a continuous number into a discrete state...

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,227 @@
+    s = sum(vec)
+    if s == 0.0:
+        n = len(vec)
+        return [1 /n for v in vec]
```
space after `/`. Also consider just using numpy:
```
s = np.sum(vec)
if s == 0:
   s = len(vec)
return np.array(vec) / s
```

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,227 @@
+    vec = [v / s for v in vec]
+    return vec
+
+def mutate_row(row, mutation_probability):
```
A doc string would be nice, along with a comment about keeping the values within [0, 1]

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,227 @@
+        def deserialize(vector):
+            matrix = []
+            for i in range(self.num_states):
+                row = vector[self.num_states * i:self.num_states * (i + 1)]
```
space after `:`

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,227 @@
+                matrix.append(row)
+            return matrix
+
+        self.transitions_C = deserialize(vector[0:self.num_states ** 2])
```
space after `:`

marcharper commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -0,0 +1,227 @@
+
+        self.transitions_C = deserialize(vector[0:self.num_states ** 2])
+        self.transitions_D = deserialize(vector[self.num_states ** 2:2 * self.num_states ** 2])
+        self.emission_probabilities = normalize_vector(vector[2 * self.num_states ** 2:2 * self.num_states ** 2 + self.num_states])
```
break up / wrap line please

marcharper commented on tests/integration/test_hmm.py:
```diff
@@ -0,0 +1,398 @@
+        population.run(generations)
+        self.assertEqual(population.generation, 4)
+
+        # Manually read from tempo file to find best strategy
```
temp or temporary

marcharper reviewed (approved):
Some minor style comments, otherwise looks good to me. Thanks for the contribution! I'll give @drvinceknight a chance to take a look before merging.

drvinceknight commented:
Once those minor requests from @marcharper are done this looks good to me. Thanks for the contribution.

gaffney2010 commented:
I'll make these changes when I can get back to my computer.

gaffney2010 commented:
I've made those changes.

drvinceknight commented on src/axelrod_dojo/archetypes/hmm.py:
```diff
@@ -28,11 +28,16 @@ def normalize_vector(vec):
 def mutate_row(row, mutation_probability):
+    """
+    Given a row of probabilities, randomly change each entry with probability
+    `mutation_probability`.  Then if changing, then change by a value randomly
```
`Then if changing, change by a value` (extra `then`)

drvinceknight replied:
> A doc string would be nice, along with a comment about keeping the values within [0, 1]

Could we have:

```
Given a row of probabilities, randomly change each entry with probability
 `mutation_probability` (a value between 0 and 1).
```

gaffney2010 replied:
No problem.

drvinceknight reviewed (changes requested):
Very minor typo: looks good. :+1:

gaffney2010 committed:
Made minor formatting fixes.

gaffney2010 committed:
Put conversion back in.

drvinceknight commented:
Looks good to me. Will give @marcharper final sign off because of the change with the `int` etc. :+1:

gaffney2010 committed:
Cleaned up docstrings.
